File: UILabel+RectForSubstring.m
//
//  UILabel+RectForSubstring.m
//  Podcat
//
//  Created by Jeanette Müller on 20.10.14.
//  Copyright (c) 2014 Jeanette Müller. All rights reserved.
//

#import "UILabel+RectForSubstring.h"
#import <CoreText/CoreText.h>

@implementation UILabel (RectForSubstring)

- (CGRect)rectForSubstringWithRange:(NSRange)range{
    
    NSTextStorage *textStorage = [[NSTextStorage alloc] initWithAttributedString:[self attributedText]];
    NSLayoutManager *layoutManager = [[NSLayoutManager alloc] init];
    [textStorage addLayoutManager:layoutManager];
    NSTextContainer *textContainer = [[NSTextContainer alloc] initWithSize:[self bounds].size];
    textContainer.lineFragmentPadding = 0;
    [layoutManager addTextContainer:textContainer];
    
    NSRange glyphRange;
    
    // Convert the range for glyphs.
    [layoutManager characterRangeForGlyphRange:range actualGlyphRange:&glyphRange];
    
    return [layoutManager boundingRectForGlyphRange:glyphRange inTextContainer:textContainer];

}

@end

File: Swift3/UILabel+RectForSubstring.swift
//
//  UILabel+RectForSubstring.swift
//  projectPhoenix
//
//  Created by Jeanette Müller on 15.12.16.
//  Copyright © 2016 Jeanette Müller. All rights reserved.
//

import UIKit

extension UILabel {
    
    func rectForSubstring(withRange range:NSRange) -> CGRect?{
        
        guard let attributedText = attributedText else { return nil }
        
        let textStorage = NSTextStorage(attributedString: attributedText)
        
        let layoutManager = NSLayoutManager()
        textStorage.addLayoutManager(layoutManager)
        
        let textContainer = NSTextContainer(size:self.bounds.size)
        textContainer.lineFragmentPadding = 0;
        layoutManager.addTextContainer(textContainer)
        
        var glyphRange:NSRange = NSRange()
        
        layoutManager.characterRange(forGlyphRange:range, actualGlyphRange:&glyphRange)
        
        return layoutManager.boundingRect(forGlyphRange:glyphRange, in:textContainer)
        
    }
}
